frontend/src/app/testQueue/QueueItem.tsx

```diff
@@ -304,23 +304,30 @@ const QueueItem: any = ({
         if (!response.data) throw Error("updateQueueItem null response");
         updateDeviceId(response.data.editQueueItem.deviceType.internalId);
         updateTestResultValue(response.data.editQueueItem.result || undefined);
-        updateDateTested(getDate(response.data.editQueueItem.dateTested));
       })
       .catch(updateMutationError);
   };
 
-  const onDeviceChange = (e: React.FormEvent<HTMLSelectElement>) => {
+  const onDeviceChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
     const deviceId = e.currentTarget.value;
     updateQueueItem({ deviceId, dateTested, result: testResultValue });
   };
 
-  const onDateTestedChange = (e: React.FormEvent<HTMLInputElement>) => {
-    const dateTested = (e.target as HTMLInputElement).value;
-
-    updateQueueItem({
-      dateTested: dateTested === "" ? undefined : dateTested,
-      result: testResultValue,
-    });
+  const onDateTestedChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const dateTested = e.target.value;
+    // If we try to round-trip the date to the server, it is too slow and the
+    // new prop clobbers what the user typed after the request continued.
+    // So, only save complete recent dates but always update locally.
+    // This will need further fixing once we go to subscriptions, ugh.
+    updateDateTested(dateTested);
+    const thisYear = new Date().getFullYear();
+    const check = new RegExp(`^(${thisYear - 1}|${thisYear})-\\d\\d-\\d\\d`);
```

Comment:
The `<input type="date" ... />` here always renders its date in `yyyy-mm-dd` format but displays it in localized format. I'm only treating the current and previous year as valid to reduce the chances of a minor typo updating the server with a bad date. I'm assuming we don't want to let people go crazy with back or forward dating but this is only checking the year. Once we get real form field validation we can narrow the date range (e.g. only back a month) and style the input to make it clear the data isn't good so we can warn before submit.